### ISSUE TITLE
[Feature Request] improve developer experience with login.microsoftonline.com in B2C tenants

### ISSUE THREAD

**Issue by prabh-62** [AzureAD/microsoft-identity-web#143]
## What do do:

When B2C is detected (presence of a policy) but the authority is login.microsoftonline.com without the tfp, we should log an error and throw an exception ArgumentException in AddMicrosoftWebApp(), so that customers know that they either have to use login.b2c.com or have tfp.

Error message:
- the configuration contains a sign-in sign-up policy and therefore your application is considered as a B2C application, however, the authority is login.microsoftonline.com. Consider either changing your authority to "https://login.microsoftonline.com/tfp/" or moving to "b2clogin.com". See https://aka.ms/ms-id-web/troubleshooting/b2c

cc: @jennyf19 
Would https://github.com/AzureAD/microsoft-identity-web/issues/168 be a duplicate?

### Documentation Related To Component:
[Microsoft.Identity.Web](https://www.nuget.org/packages/Microsoft.Identity.Web/0.1.1-preview) nuget package (version 0.1.1-preview)

### Please check those that apply

- [ ] typo
- [ ] documentation doesn't exist
- [x] documentation needs clarification
- [x] error(s) in example
- [x] needs example

### Description Of The Issue
- `> dotnet --version`
			`3.1.201`
- `> git clone https://github.com/AzureAD/microsoft-identity-web.git`
- `> cd microsoft-identity-web/ProjectTemplates`
- `> dotnet pack AspNetCoreMicrosoftIdentityWebProjectTemplates.csproj`
- `> cd bin/Debug`
- `> dotnet new --install Microsoft.Identity.Web.ProjectTemplates.0.1.0.nupkg`

Verify if the templates are available
- `> dotnet new`
<img width="1283" alt="webapi2" src="https://user-images.githubusercontent.com/13006107/81031346-f876f300-8e59-11ea-8d39-c95b199f7dd5.png">

Now, let's create a new ASP.NET Core web api project
- Navigate to a folder where we will create new project
- `> dotnet new webapi2 --auth Singleorg -n WeatherStation`
- `> cd WeatherStation`
- Update `appsettings.json` file (image has fake data)
<img width="546" alt="app_settings" src="https://user-images.githubusercontent.com/13006107/81031436-40961580-8e5a-11ea-9b67-e5ae14d359fc.png">

- `> dotnet run`

- `> curl -i http://localhost:5000/weatherforecast -H  "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI2NiIsImtpZCI6Ilg1ZVhrNHh5b2pORnVtMWtsMll0djhkbE5QNC1jNTdkTzZRR1RWQndhTmsifQ.eyJpc3MiOiJodHRwczovL2xvZ2luLm1pY3Jvc29mdG9ubGluZS5jb20vNzNiOWUwZjgtNDY1Yy00MzY1LTkzNGUtMTM4ZmUxYjEyNDQ4L3YyLjAvIiwiZXhwIjoxNTg4NjQ2NTIyLCJuYmYi9wbWVudCIsImZhbWlseV9uYW1lIjoiR3JvdXAiLCJuYW1lIjoiRGV2ZWxvcG1lbnQgR3JvdXAiLCJuZXdVc3VyIjpmYWxzZSwiZXh0ZW5zaW9uX0NvbXBhbnkiOiJNSUNST0RFQSIsImV4dGVuc2lvbl9QaG9uZSI6IjU1NTU1NTU1NTUiLCJlbWFpbHMiOlsiZGV2ZWxvcG1lbnRncm91cEBtaWNyb2RlYS5jb20iXSwidGZwIjoiQjJDXzFfTG9hZFBhbERldiIsImF6cCI6IjliNzBjZTY4LTZlOWEtNGJiZC1iYWU3LWVhZGY5YjFkN2IxMSIsInZlciI6IjEuMCIsImlhdCI6MTU4ODYzOTMyMn0.hVAD5SZ4hIsyYdZgKPT0LBzHP3Ud5aU8vHXWQBcCMMOcjb5ZApiZSjzI7fdEQHuesudQtgwZumui-1a_XIV6v6jls5I_SlCr-h5bKJwa1VAW7_oKmKVxEjqt60dVJU8LIizySXimNXpS8W-YUHz0HBptE1vHndwadOT2OvB2ZOOHhNUnpNBdxaCYR-0TdSeH2ZnpXs6mphzxyRdD8-Bt7BB4FJZUNH63HpsJ3cV7aO08FrJ0jkveIdwcFy2WZbW-i1B8NWaWgPOpyx3DTWm3UCfJsLmVy21d6sK8LBL-vRaBfiSIfR9I1L2W_hB9U-TQMaTwQkAuXh4cNmg2u7GT8P"`
<img width="1421" alt="signature_keys_not_found" src="https://user-images.githubusercontent.com/13006107/81031492-69b6a600-8e5a-11ea-9b26-0c71379632ec.png">

I get an error message
`WWW-Authenticate: Bearer error="invalid_token", error_description="The signature key was not found"`

**Comment by prabh-62** [comment]
I am able to retrieve list of users using `GraphServiceClient`
<img width="1145" alt="graphServiceAccount" src="https://user-images.githubusercontent.com/13006107/81032086-73410d80-8e5c-11ea-9c46-177163509b38.png">

Nugets in project:
```xml
    <PackageReference Include="Microsoft.Graph" Version="3.4.0" />
    <PackageReference Include="Microsoft.Graph.Auth" Version="1.0.0-preview.4" />
    <PackageReference Include="Microsoft.Identity.Client" Version="4.12.0" />
    <PackageReference Include="Microsoft.Identity.Web" Version="0.1.1-preview" />
```

**Comment by jmprieur** [comment]
Hello @prabh-62 : did you look at this sample: https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/tree/master/2-WebApp-graph-user/2-1-Call-MSGraph

**Comment by prabh-62** [comment]
Yes, I looked through multiple projects in that git repository. I tried the same approach as [4-WebApp-your-API/4-1-MyOrg](https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/master/4-WebApp-your-API/4-1-MyOrg/TodoListService/Startup.cs#L34). 
<img width="779" alt="startup_protected_web_api" src="https://user-images.githubusercontent.com/13006107/81106327-65829b00-8ee3-11ea-8ef7-2199578c03e4.png">
<img width="500" alt="authorize" src="https://user-images.githubusercontent.com/13006107/81106356-6fa49980-8ee3-11ea-979b-c320b40903f6.png">



I still get the same error.
```
HTTP/1.1 401 Unauthorized
Date: Tue, 05 May 2020 19:08:39 GMT
Server: Kestrel
Content-Length: 0
WWW-Authenticate: Bearer error="invalid_token", error_description="The signature key was not found"
```
<img width="1385" alt="signature_key" src="https://user-images.githubusercontent.com/13006107/81106182-22282c80-8ee3-11ea-8fff-6ad20426a9c8.png">


Approach `2-1-Call-MSGraph` uses [Microsoft.Identity.Web.UI](https://www.nuget.org/packages/Microsoft.Identity.Web.UI) and I am building an API so I cannot use that example.

**Comment by jmprieur** [comment]
IF you are building a Web API you need to look at https://github.com/Azure-Samples/active-directory-dotnet-native-aspnetcore-v2
Would your API require JWE? (encrypted tokens) ?

**Comment by prabh-62** [comment]
Access tokens in the API are normal JWTs.
I copied the exact approach from [3.-Web-api-call-Microsoft-graph-for-personal-accounts](https://github.com/Azure-Samples/active-directory-dotnet-native-aspnetcore-v2/blob/master/3.-Web-api-call-Microsoft-graph-for-personal-accounts/TodoListService/Startup.cs#L26)
```csharp
			services.AddProtectedWebApi(Configuration)
                    .AddProtectedWebApiCallsProtectedWebApi(Configuration)
                    .AddInMemoryTokenCaches();
```
<img width="1404" alt="signature_key_not_found" src="https://user-images.githubusercontent.com/13006107/81111802-a1b9f980-8eeb-11ea-88b0-224e59e4dc47.png">

**Comment by jmprieur** [comment]
Did you try this one?
https://github.com/Azure-Samples/active-directory-dotnet-native-aspnetcore-v2/tree/master/2.%20Web%20API%20now%20calls%20Microsoft%20Graph

**Comment by prabh-62** [comment]
These are the steps I followed
- `> git clone https://github.com/Azure-Samples/active-directory-dotnet-native-aspnetcore-v2.git`
- `> cd active-directory-dotnet-native-aspnetcore-v2/2.\ Web\ API\ now\ calls\ Microsoft\ Graph/TodoListService/`
-  Update credentials in `appsettings.json`
- `> dotnet run`
- Make a post call
<img width="1405" alt="curl_post_call" src="https://user-images.githubusercontent.com/13006107/81114774-a9c86800-8ef0-11ea-9139-9909b3a07830.png">
- I see exceptions
<img width="1405" alt="PII_signature_key" src="https://user-images.githubusercontent.com/13006107/81114795-b482fd00-8ef0-11ea-9a74-467552254f04.png">

**Comment by prabh-62** [comment]
I was able to get the AzureAdB2C Authentication working with the ASP.NET Core.
```csharp
 public void ConfigureServices(IServiceCollection services)
        {
            var discoveryPoint = "https://login.microsoftonline.com/{DirectoryID}/v2.0/.well-known/openid-configuration?p={Policy}";
            var configManager =
                    new ConfigurationManager<OpenIdConnectConfiguration>(
                        discoveryPoint,
                        new OpenIdConnectConfigurationRetriever()
                    );

            var config = configManager.GetConfigurationAsync(CancellationToken.None).GetAwaiter().GetResult();

            services.AddProtectedWebApi(options =>
            {
                Configuration.Bind("AzureAdB2C", options);
                options.IncludeErrorDetails = true;

                options.TokenValidationParameters = new TokenValidationParameters
                {
                    RequireSignedTokens = true,
                    ValidIssuer = "https://login.microsoftonline.com/{DirectoryId}/v2.0/",
                    ValidAudience = "{ClientId}",
                    ValidateAudience = true,
                    ValidateIssuer = true,
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKeys = config.SigningKeys,
                    ValidateLifetime = true
                };

                options.Events = new JwtBearerEvents
                {
                    OnAuthenticationFailed = context =>
                    {
                        context.Response.OnStarting(async () =>
                        {
                            context.NoResult();
                            context.Response.ContentType = "application/json";
                            byte[] bytes = Encoding.ASCII.GetBytes(context.Exception.Message);
                            await context.Response.Body.WriteAsync(bytes);
                            context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
                        });

                        return Task.CompletedTask;
                    }
                };
            }, options => Configuration.Bind("AzureAdB2C", options));
        }
``` 

However, there is one small hurdle. I had to comment certain code in file `src/Microsoft.Identity.Web/WebApiAuthenticationBuilderExtensions.cs` of `Microsoft.Identity.Web` library 
```csharp
                options.Events.OnTokenValidated = async context =>
                {
                    // This check is required to ensure that the Web API only accepts tokens from tenants where it has been consented and provisioned.
                    // if (!context.Principal.Claims.Any(x => x.Type == ClaimConstants.Scope)
                    // && !context.Principal.Claims.Any(y => y.Type == ClaimConstants.Scp)
                    // && !context.Principal.Claims.Any(y => y.Type == ClaimConstants.Roles)
                    // && !context.Principal.Claims.Any(y => y.Type == ClaimConstants.Role))
                    // {
                    //     throw new UnauthorizedAccessException("Neither scope or roles claim was found in the bearer token.");
                    // }

                    await tokenValidatedHandler(context).ConfigureAwait(false);
                };
```
Exact Line link: https://github.com/AzureAD/microsoft-identity-web/blob/master/src/Microsoft.Identity.Web/WebApiAuthenticationBuilderExtensions.cs#L132
Is it possible to not throw exceptions in certain situations?

**Comment by jmprieur** [comment]
Oh I see you were trying to use B2C. the issue is that you have used 

`dotnet new webapi2 --auth Singleorg -n WeatherStation`

and you should really use

`dotnet new webapi2 --auth IndividualB2C-n WeatherStation`

**Comment by prabh-62** [comment]
This morning, I tried again

- `> dotnet new webapi2 --auth IndividualB2C -n WeatherStation`
- Updated `AzureAdB2C` section in `appsettings.json`
- Disabled HTTPS redirection
	```csharp
		public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
                IdentityModelEventSource.ShowPII = true;
            }

            // app.UseHttpsRedirection();
	```
- `curl -i http://localhost:5000/weatherforecast -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6Ilg1ZVhrNHh5b2pORnVtMWtsMll0djhkbE5QNC1jNTdkTzZRR1RWQndhTmsifQ.eyJpc3MiOiJodHRwczovL2xvZ2luLm1pY3Jvc29mdG9ubGluZS5jb20vNzNiOWUwZjgtNDY1Yy00MzY1LTkzNGUtMTM4ZmUxYjEyNDQ4L3YyLjAvIiwiZXhwIjoxNTg5MjEwMzcyLCJuYmYiOjE1ODkyMDMxNzIsImF1ZCI6IjliNzBjZTY4LTZlOWEtNGJiZC1iYWU3LWVhZGY5YjFkN2IxMSIsImlkcCI6IkxvY2FsQWNjb3VudCIsIm9pZCI6ImFjMWMwYzJmLWJiNjgtNDU1YS1hMTcxLTdiMjhjNDBlNTQwZCIsInN1YiI6ImFjMWMwYzJmLWJiNjgtNDU1YS1hMTcxLTdiMjhjNDBlNTQwZCIsImdpdmVuX25hbWUiOiJEZXZlbG9wbWVudCIsImZhbWlseV9uYW1lIjoiR3JvdXAiLCJuYW1lIjoiRGV2ZWxvcG1lbnQgR3JvdXAiLCJuZXdVc2VyIjpmYWxzZSwiZXh0ZW5zaW9uX0NvbXBhbnkiOiJNSUNST0RFQSIsImV4dGVuc2lvbl9QaG9uZSI6IjU1NTU1NTU1NTUiLCJlbWFpbHMiOlsiZGV2ZWxvcG1lbnRncm91cEBtaWNyb2RlYS5jb20iXSwidGZwIjoiQjJDXzFfTG9hZFBhbERldiIsImF6cCI6IjliNzBjZTY4LTZlOWEtNGJiZC1iYWU3LWVhZGY5YjFkN2IxMSIsInZlciI6IjEuMCIsImlhdCI6MTU4OTIwMzE3Mn0.JIQkmQ_3MTwpt3WheWERdY7YEDx48Yef3kitiUW-2tnThOACqblWGxBSek8HarAyvahlVr6jnENuipvW_px6nJXu0EWZMJTDQ9wkeIPdVNlbws3iwvh2hctTFUAztVWo_noSdFRqeb_U95sTLsCTEu3ck4mLBYOxuCsHhKlyeui_8n6R57J4-B0Jjk_BUsadsS_0GHKHJeAaJ14M2QEtertUZIgQQIqFb07OwrB_PIvyTZxaofPAwKEkvWwwMtns-xjCpQNlje41py0eUl6cfDjIUpOWyl2alk5ASjrqaZK2sYx3JDkSUDun33UOh9VIcog_JBQ7HVREF97bCVnp7w"`

And I get an exception 
<img width="1187" alt="pII_error" src="https://user-images.githubusercontent.com/13006107/81571284-b83ed580-936f-11ea-80c2-8901141cba9f.png">

**Comment by jennyf19** [comment]
@prabh-62 What is the instance and domain value in appsettings.json for the b2c tenant?

**Comment by prabh-62** [comment]
This is how my `appsettings.json` looks like
```json
{
  "AzureAdB2C": {
    "Instance": "https://login.microsoftonline.com",
    "ClientId": "Guid",
    "B2cAppId": "Guid",
    "DirectoryId": "Guid",
    "Domain": "MicrodeaDev.onmicrosoft.com",
    "SignUpSignInPolicyId": "B2C_1_SignupSignin"
  },
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Information"
    }
  },
  "AllowedHosts": "*"
}
```

**Comment by jennyf19** [comment]
@prabh-62 thank you.
If you're using B2C, you have to include the policy in order to get the correct metadata and endpoints.

So, if you're using login.microsoftonline.com + policy (which is being deprecated as a b2c endpoint), you need to include the policy: 
`https://login.microsoftonline.com/tfp/{domain}/{policy}/v2.0/.well-known/openid-configuration`

Did you create an app in your B2C tenant? You cannot mix the AzureAd settings and B2C settings, they are separate authorization servers at the moment, if i'm understanding your scenario correctly.

You should also be using [*.b2clogin.com](https://docs.microsoft.com/en-us/azure/active-directory-b2c/b2clogin) now as the tenant. Let me know if this helps.

**Comment by prabh-62** [comment]
Thank you for the help. We will be migrating to *.b2clogin.com soon. I updated the `appsettings.json`
```json
{
    "AzureAdB2C": {
      "Instance": "https://login.microsoftonline.com/tfp/",
      "ClientId": "Guid",
      "B2cAppId": "Guid",
      "DirectoryId": "Guid",
      "Domain": "MicrodeaDev.onmicrosoft.com",
      "SignUpSignInPolicyId": "B2C_1_SignupSignin"
    },
    "Logging": {
      "LogLevel": {
        "Default": "Information",
        "Microsoft": "Warning",
        "Microsoft.Hosting.Lifetime": "Information"
      }
    },
    "AllowedHosts": "*"
  }
```

Also, I had to comment out the [following lines in the library](https://github.com/AzureAD/microsoft-identity-web/blob/master/src/Microsoft.Identity.Web/WebApiAuthenticationBuilderExtensions.cs#L138)
<img width="1110" alt="scope_check_identity" src="https://user-images.githubusercontent.com/13006107/81617077-bc401700-93b2-11ea-8253-8ca34337d2c1.png">



The authentication is working now with the following startup code.
<img width="679" alt="startup_code" src="https://user-images.githubusercontent.com/13006107/81617140-e1cd2080-93b2-11ea-9775-adcdf5458d44.png">

🤗

**Comment by jennyf19** [comment]
@prabh-62 thanks for the quick reply.

@jmprieur how would you like to proceed w/handling this use case (b2c w/login.microsoftonline.com)?

**Comment by jmprieur** [comment]
Thanks @jennyf19 
Do we want to throw in  Microsoft.Identity.Web if we detect B2C (a user flow) and authority is login.microsoftonline.com without the tfp ? We could have a meaningful exception advising to use b2cloging.com? what do you think?

**Comment by bgavrilMS** [comment]
#supportability

**Comment by jennyf19** [comment]
@jmprieur let's discuss.

**Comment by jennyf19** [comment]
@prabh-62 i updated the title, as i believe this was the root cause. will make it easier for us to track. hope you don't mind. :)

**Comment by jmprieur** [comment]
One of the issues is for a b2C web app that calls a web API, if the developer provides the tfp in the authority in the appsettings.json, the sign-in part of the Oauth code flow works, but Microsoft.Identity.Web adds a second /tfp in the authority, and MSAL fails.

```JSon
{
  "AzureAdB2C": {
    "Instance": "https://login.microsoftonline.com/tfp/",
```

Possible work around for this case (and only this case) would be to add the following line after
https://github.com/AzureAD/microsoft-identity-web/blob/4ab0da4c93ef554fab397cb4f1a873ff62d392c3/src/Microsoft.Identity.Web/TokenAcquisition.cs#L152

```CSharp
   // Consider the case where the authority in the config ends with /tfp
   authority = authority.Replace("/tfp/tfp", "/tfp");
```

This is an issue because the ASP.NET Core templates do that.
Proposal:
1. apply this work around to Microsoft.Identity.Web
2. Have the ASP.NET Core templates changed to use b2clogin.com instead of https://login.microsoftonline.com/tfp/

**Comment by jennyf19** [comment]
Included in [1.4 Release](https://github.com/AzureAD/microsoft-identity-web/releases/tag/1.4.0).
